### PR TITLE
Update terminal width each time we print a display

### LINKED ISSUE #4732: The new progress bars do not react well to the changes in tty width
My strongest dislike of the progress bars is how they mishandle the terminal width changes all the time. In case of cargo, there’s no handling at all.

For example see the attached image of what happens when the terminal is wider at first and then its width is decreased:

Image with a lot of repeating progress bars

This is a surprisingly common problem for me, who tends to change tiling mode, and thus terminal widths, fairly often.

### PR DESCRIPTION
**Pull request by alexcrichton**
This should help accomodate terminals that are resized while a progress bar is
being displayed. Additionally this shouldn't come with too much of a performance
impact because we're already heavily throttling printing of a progress update!

Closes #4732

### CODE REVIEW

File: src/cargo/util/progress.rs [hunk 1]
```diff
@@ -127,12 +127,25 @@ impl<'cfg> State<'cfg> {
...
+        self.try_update_max_width();
         if let Some(string) = self.format.progress_status(cur, max, msg) {
             self.config.shell().status_header(&self.name)?;
             write!(self.config.shell().err(), "{}\r", string)?;
         }
         Ok(())
     }
+
+    fn clear(&mut self) {
+        self.try_update_max_width();
```
Comment (matklad) [inline src/cargo/util/progress.rs]: Shouldn't we use *previous* width here? I am thinking about the following sequence:

1) we print 80col line
2) user resizes terminal to 40 col, so the line wraps 
3) clear rubs-out only 40 cols, so we get both a wrapping and a tail of old line on the screen.

Comment (alexcrichton) [reply src/cargo/util/progress.rs]: Nah at least in my testing the current width is what you always want I think. The behavior I saw is that if you had a 10-character wide window with text that looked like:

```
0123456789
```

then if you resize it to 9 characters you get:

```
        0
123456789
```

In that when you make the window smaller it pushes the contents at the start of the line to the previous line. It certainly makes everything look funky, but your cursor is always placed at the very last line and printing the terminal's width always clears just what the terminal has

File: src/cargo/util/progress.rs [hunk 2]
```diff
@@ -196,24 +209,23 @@ impl Format {
...
 }
 
 impl<'cfg> Drop for State<'cfg> {
     fn drop(&mut self) {
-        clear(self.format.max_width, self.config);
+        self.clear();
     }
 }
 
 #[test]
```
Comment (kennytm) [inline src/cargo/util/progress.rs]: Perhaps add a test case where the `max_width` is mutated 😃

**Comment by alexcrichton** [comment]
ping r? @matklad

File: src/cargo/util/progress.rs (hunk 2 above)
Comment (alexcrichton) [reply src/cargo/util/progress.rs]: On the `Format` struct though the methods are all immutable so it may not be testing much :(

**Pull request merged**